bundles/org.openhab.ui/web/src/components/widgets/widget-mixin.js

```diff
@@ -144,6 +145,16 @@ export default {
         return value
       }
     },
+    getPageContentRect () {
+      const page = document.getElementById('page') !== null ? document.getElementById('page').getBoundingClientRect() : null
+      const navbar = document.getElementById('page-navbar') !== null ? document.getElementById('page-navbar').getBoundingClientRect() : null
+      const toolbar = document.getElementById('page-toolbar') !== null ? document.getElementById('page-toolbar').getBoundingClientRect() : null
```

Comment:
- See remark below about uniqueness of pages. The home page also has a navbar (potentially with a dynamic height) & bottom toolbar which is not taken into account here. Look for the element with class `view` or `page-current` instead. Then the page area excluding navbar & toolbar is the descendent of `page-current` with the `page-content` class.
- When the page is displayed in a popup/modal those values don't reflect the popup's size, not sure if that's what you want? Pages in modals also have `page-content` which would be the area of the modal
- These values don't update when the window is resized or the screen orientation changed, if it's not an issue then fine, otherwise a `calc(100vh - var(--f7-navbar-height))` or similar in CSS would be more elegant and better practice than getting pixel values for positioning

Reply:
> Look for the element with class view or page-current instead. Then the page area excluding navbar & toolbar is the descendent of page-current with the page-content class.

Thanks for the tip, I've implemented this :+1: .

I am now selecting/getting the elements by their respective class name and I am also subtracting the padding from width and height.

> These values don't update when the window is resized or the screen orientation changed,

I am aware that these values are not dynamic, but as long as I find no pure CSS way to do comparisons like the following, I need those value inside the JS expressions:

```yaml
    - component: oh-image
      config:
        url: image-url
        refreshInterval: 250
        lazy: false
        style:
          margin-left: "=screen.width <= (screen.height * (4/3)) ? '0' : 'calc(calc(100% - calc(96vh - var(--f7-navbar-height) - var(--f7-toolbar-height)) * (4/3)) / 2)'"
          border-radius: var(--f7-card-expandable-border-radius)
          width: "=screen.width <= (screen.height * (4/3)) ? '100%' : 'calc(calc(96vh - var(--f7-navbar-height) - var(--f7-toolbar-height)) * (4/3))'"
```

> otherwise a calc(100vh - var(--f7-navbar-height)) or similar in CSS would be more elegant and better practice than getting pixel values for positioning

I am using CSS calc for the exact positioning, getting the pixel values is only important for making the decision which CSS statement should be passed.

> When the page is displayed in a popup/modal those values don't reflect the popup's size, not sure if that's what you want? Pages in modals also have page-content which would be the area of the modal

That's the problem I have no solution for, but personally I am not using popups in combination with getting the pixel values of width and height.